### PR TITLE
Invariant improvements

### PR DESCRIPTION
**Pull request by MikeHathaway**
# Description of change
## High level
* Various improvements to the invariant tests in GrantFund, including all remaining TODOs.
  * Flip a coin to decide whether to cast a negative funding vote
  * Add `P1` invariant check
  * Add `CS7` invariant check
  * Improve `DR5` invariant check
  * Improve `ES2` invariant check
  * Add `Logger` class to centralize all logging related logic
  * Add `test-invariant.sh` script and support for env variables in Invariant tests
  * Update Makefile

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,10 +15,11 @@ install :; git submodule update --init --recursive
 build  :; forge clean && forge build --optimize --optimizer-runs 1000000
 
 # Tests
-tests   :; forge clean && forge test --mt test --optimize --optimizer-runs 1000000 -v # --ffi # enable if you need the `ffi` cheat code on HEVM
-test-with-gas-report   :; forge clean && forge build && forge test --mt test --optimize --optimizer-runs 1000000 -v --gas-report # --ffi # enable if you need the `ffi` cheat code on HEVM
-test-invariant		 :; forge clean && forge t --mt invariant
-coverage   :; forge coverage
+tests   			   				 :; forge clean && forge test --mt test --optimize --optimizer-runs 1000000 -v # --ffi # enable if you need the `ffi` cheat code on HEVM
+test-with-gas-report   				 :; forge clean && forge build && forge test --mt test --optimize --optimizer-runs 1000000 -v --gas-report # --ffi # enable if you need the `ffi` cheat code on HEVM
+test-invariant-all-scenarios		 :; forge clean && forge t --mt invariant
```
Comment (EdNoepel) [inline Makefile]: Old target name seems cleaner, or could name `test-invariant-all` for consistency with _contracts_ repo.

File: Makefile [hunk 2]
```diff
@@ -15,10 +15,11 @@ install :; git submodule update --init --recursive
...
+test-invariant-multiple-distribution :; forge clean && ./test/invariants/test-invariant.sh MultipleDistribution 25 200
+coverage   			   				 :; forge coverage
```
Comment (EdNoepel) [inline Makefile:23]: nit: Alignment of targets and instructions was weird before, but seems worse now.  Is possible to line up the `:;`?

File: foundry.toml [hunk 3]
```diff
@@ -9,12 +9,13 @@ block_number        = 16_295_000
 runs = 150
 
 [invariant]
-runs           = 3     # The number of calls to make in the invariant tests
+runs           = 3    # The number of calls to make in the invariant tests
 depth          = 10000   # The number of times to run the invariant tests
```
Comment (EdNoepel) [inline foundry.toml]: nit: comment alignment on lines 12 and 13

File: test/invariants/INVARIANTS.md [hunk 4]
```diff
@@ -22,13 +22,11 @@
     - **SS4**: Screening vote's cast can only be positive.
     - **SS5**: Screening votes can only be cast on a proposal in it's distribution period's screening stage.
     - **SS6**: For every proposal, it is included in the top 10 list if, and only if, it has as many or more votes as the last member of the top ten list (typically the 10th of course, but it may be shorter than ten proposals).
-    - **SS7**: Screening vote's on a proposal should cause addition to the topTenProposals if no proposal has been added yet.
-    <!-- TODO: move these into separate 'Proposal invariants:' section -->
-    - **SS8**: A proposal should never receive more vote than the Ajna token supply.
+    - **SS7**: Screening vote's on a proposal should cause addition to the topTenProposals if no proposal has been added yet
```
Comment (EdNoepel) [inline test/invariants/INVARIANTS.md]: nit: spelling `votes`; apostrophe implies possession

File: test/invariants/base/StandardTestBase.sol [hunk 5]
```diff
@@ -26,11 +29,15 @@ contract StandardTestBase is DistributionPeriodInvariants, FinalizeInvariants, F
             payable(address(_grantFund)),
             address(_ajna),
             _tokenDeployer,
-            NUM_ACTORS,
+            vm.envOr("NUM_ACTORS", NUM_ACTORS),
+            vm.envOr("NUM_PROPOSALS", NUM_PROPOSALS),
```
Comment (EdNoepel) [inline test/invariants/base/StandardTestBase.sol:34]: Glad these can be loaded from environment.  Recommend documenting by updating `.env.example`.

File: test/invariants/test-invariant.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+NUM_PROPOSALS=$3
+PER_ADDRESS_TOKEN_REQ_CAP=$4
+
+echo "Exporting environment variables"
+
+# export environment variables
+export SCENARIO=$SCENARIO
+export NUM_ACTORS=$NUM_ACTORS
+export NUM_PROPOSALS=$NUM_PROPOSALS
+export PER_ADDRESS_TOKEN_REQ_CAP=$PER_ADDRESS_TOKEN_REQ_CAP
```
Comment (EdNoepel) [inline test/invariants/test-invariant.sh]: You could eliminate lines 4-11 and replace 13-16 with:
```
export SCENARIO=${1:-${SCENARIO}}
export NUM_ACTORS=${2:-${NUM_ACTORS}}
export NUM_PROPOSALS=${3:-${NUM_PROPOSALS}}
export PER_ADDRESS_TOKEN_REQ_CAP=${4:-${PER_ADDRESS_TOKEN_REQ_CAP}}
```
The `${n:-$VAL}` means use the nth argument, or default to _$VAL_ if not provided.

**Review by EdNoepel — APPROVED** [review]
Looks good; just some cosmetic things.

File: Makefile (hunk 2 above)
Comment (MikeHathaway) [reply Makefile:23]: Yeah I was looking at that - I had changed it to the alignment that was clean locally. Will reinvestigate

File: test/invariants/INVARIANTS.md (hunk 4 above)
Comment (MikeHathaway) [reply test/invariants/INVARIANTS.md]: Fixed

File: Makefile (hunk 1 above)
Comment (MikeHathaway) [reply Makefile]: Renamed to `test-invariant-all`

File: test/invariants/test-invariant.sh (hunk 6 above)
Comment (MikeHathaway) [reply test/invariants/test-invariant.sh]: Awesome, that's much cleaner. Updated accordingly.

**Review by prat-gpt — APPROVED** [review]
LGTM!

**Pull request merged**